homeassistant/components/rainbird/sensor.py

```diff
@@ -3,58 +3,53 @@
 
 import logging
 
-from pyrainbird import RainbirdController
-
 from homeassistant.components.sensor import SensorEntity, SensorEntityDescription
 from homeassistant.core import HomeAssistant
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
-from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
-
-from . import (
-    DATA_RAINBIRD,
-    RAINBIRD_CONTROLLER,
-    SENSOR_TYPE_RAINDELAY,
-    SENSOR_TYPE_RAINSENSOR,
-    SENSOR_TYPES,
+from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType, StateType
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
 )
 
+from . import SENSOR_TYPES
+
 _LOGGER = logging.getLogger(__name__)
 
 
-def setup_platform(
+async def async_setup_platform(
     hass: HomeAssistant,
     config: ConfigType,
-    add_entities: AddEntitiesCallback,
+    async_add_entities: AddEntitiesCallback,
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up a Rain Bird sensor."""
 
     if discovery_info is None:
         return
 
-    controller = hass.data[DATA_RAINBIRD][discovery_info[RAINBIRD_CONTROLLER]]
-    add_entities(
-        [RainBirdSensor(controller, description) for description in SENSOR_TYPES],
+    async_add_entities(
+        [
+            RainBirdSensor(discovery_info[description.key], description)
+            for description in SENSOR_TYPES
+        ],
         True,
```

Comment:
Should this still call updated_before_add=True or can you drop that and write as a generator expression?

```python
    async_add_entities(
        RainBirdSensor(discovery_info[description.key], description)
        for description in SENSOR_TYPES
    )
```

Reply:
It still needs `True`.  Perhaps because the coordinator is not calling `async_config_entry_first_refresh`? Generally, i'm unsure how update coordinators are supposed to work and feel like I must be using them wrong.

Reply:
When its set to `True` its going to call `async_update` when adding each entity and since they are `CoordinatorEntity` each one of them will call https://github.com/home-assistant/core/blob/b8ce10f9a148b48e9046253b7a277794355b4e02/homeassistant/helpers/update_coordinator.py#L392

Its probably ok because there is a Debounce https://github.com/home-assistant/core/blob/b8ce10f9a148b48e9046253b7a277794355b4e02/homeassistant/helpers/update_coordinator.py#L82 but its going to generate a lot of unneeded tasks.  

Ideally you want to refresh it before it gets to here but because its using the legacy `async_setup_platform` you are probably stuck with `True` here since if you do it in `async_setup` its not possible to raise `PlatformNotReady` there.  

Since you plan on converting it to a config entry and this small inefficiency is going to go away anyways it seems just fine as an intermediate step vs trying to cram it all in one PR